nav2_costmap_2d/plugins/semantic_segmentation_layer.cpp

```diff
@@ -0,0 +1,304 @@
+    expected_update_rate;
+
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  declareParameter("combination_method", rclcpp::ParameterValue(1));
+  declareParameter("publish_debug_topics", rclcpp::ParameterValue(false));
+  declareParameter("max_obstacle_distance", rclcpp::ParameterValue(5.0));
+  declareParameter("min_obstacle_distance", rclcpp::ParameterValue(0.3));
+  declareParameter("segmentation_topic", rclcpp::ParameterValue(""));
+  declareParameter("pointcloud_topic", rclcpp::ParameterValue(""));
+  declareParameter("observation_persistence", rclcpp::ParameterValue(0.0));
+  declareParameter("expected_update_rate", rclcpp::ParameterValue(0.0));
+  declareParameter("class_types", rclcpp::ParameterValue(std::vector<std::string>({})));
+
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+  node->get_parameter(name_ + "." + "combination_method", combination_method_);
+  node->get_parameter(name_ + "." + "publish_debug_topics", debug_topics_);
+  node->get_parameter(name_ + "." + "max_obstacle_distance", max_obstacle_distance);
+  node->get_parameter(name_ + "." + "min_obstacle_distance", min_obstacle_distance);
+  node->get_parameter(name_ + "." + "segmentation_topic", segmentation_topic);
+  node->get_parameter(name_ + "." + "pointcloud_topic", pointcloud_topic);
+  node->get_parameter(name_ + "." + "observation_persistence", observation_keep_time);
+  node->get_parameter(name_ + "." + "sensor_frame", sensor_frame);
+  node->get_parameter(name_ + "." + "expected_update_rate", expected_update_rate);
+  node->get_parameter("transform_tolerance", transform_tolerance);
+  node->get_parameter(name_ + "." + "class_types", class_types_string);
+  if (class_types_string.empty())
+  {
+    RCLCPP_ERROR(logger_, "no class types defined. Segmentation plugin cannot work this way");
+    exit(-1);
+  }
+
+  for (auto& source : class_types_string)
+  {
+    // get the cost for each class
+    std::vector<std::string> classes_ids;
+    uint8_t cost;
+    declareParameter(source + ".classes", rclcpp::ParameterValue(std::vector<std::string>({})));
+    declareParameter(source + ".cost", rclcpp::ParameterValue(0));
+    node->get_parameter(name_ + "." + source + ".classes", classes_ids);
+    if (classes_ids.empty())
+    {
+      RCLCPP_ERROR(logger_, "no classes defined on type %s", source.c_str());
+      continue;
+    }
+    node->get_parameter(name_ + "." + source + ".cost", cost);
+    for (auto& class_id : classes_ids)
+    {
+      class_map_.insert(std::pair<std::string, uint8_t>(class_id, cost));
+    }
+  }
+
+  if (class_map_.empty())
+  {
+    RCLCPP_ERROR(logger_, "No classes defined. Segmentation plugin cannot work this way");
+    exit(-1);
+  }
+
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+  rolling_window_ = layered_costmap_->isRolling();
+
+  matchSize();
+
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_sensor_data;
+
+  semantic_segmentation_sub_ =
```

Comment:
Subs should use the provided callback group https://github.com/ros-planning/navigation2/blob/985fd7e61a9701f130c41fe5e95e009847465ad2/nav2_costmap_2d/plugins/obstacle_layer.cpp#L123 so that it runs on another thread